js/gdm/loginDialog.js

```diff
@@ -688,12 +711,105 @@ const LoginDialog = new Lang.Class({
+            this._customerSupportEmail = keyFile.get_locale_string('Customer Support', 'Email', null);
+            this._customerSupportPhoneNumber = keyFile.get_locale_string('Customer Support', 'Phone', null);
+        } catch (e) {
+            logError(e, 'Failed to read customer support data');
+            return false;
+        }
+
+        return this._customerSupportEmail && this._customerSupportPhoneNumber;
+    },
+
+    _maybeShowPasswordResetButton: function() {
+        if (!this._customerSupportPhoneNumber || !this._customerSupportEmail) {
+            if (!this._readCustomerSupportData())
+                return;
+        }
+
+        // There's got to be a better way to get our pid...?
+        let credentials = new Gio.Credentials();
+        let pid = credentials.get_unix_pid();
+
+        // accountsservice provides no async API, and unconditionally informs
+        // polkit that interactive authorization is permissible. If interactive
+        // authorization is attempted on the login screen during the call to
+        // set_password_mode, it will hang forever. Ensure the password reset
+        // button is hidden in this case. Besides, it's stupid to prompt for a
+        // password in order to perform password reset.
+        Polkit.Permission.new('org.freedesktop.accounts.user-administration',
+                              Polkit.UnixProcess.new_for_owner(pid, 0, -1),
+                              null,
+                              Lang.bind(this, function (obj, result) {
+                                  try {
+                                      let permission = Polkit.Permission.new_finish(result);
+                                      if (permission.get_allowed())
+                                          this._passwordResetButton.show();
+                                  } catch(e) {
+                                      logError(e, 'Failed to determine if password reset is allowed');
+                                  }
+                                }));
+    },
+
+    _generateResetCode: function() {
+        // Note: These are not secure random numbers. Doesn't matter. The
+        // mechanism to convert a reset code to unlock code is well-known, so
+        // who cares how random the reset code is?
+        let resetCode = '';
+        for (let n = 0; n < _RESET_CODE_LENGTH; n++)
+            resetCode = '%s%d'.format(resetCode, GLib.random_int_range(0, 10));
+        return resetCode;
+    },
+
+    _computeUnlockCode: function(resetCode) {
+        let checksum = new GLib.Checksum(GLib.ChecksumType.MD5);
+        checksum.update(ByteArray.fromString(resetCode));
+
+        let unlockCode = checksum.get_string();
+        unlockCode = unlockCode.replace(/\D/g, '');
+        unlockCode = unlockCode.slice(0, _RESET_CODE_LENGTH);
+
+        while (unlockCode.length < _RESET_CODE_LENGTH)
+            unlockCode += '0';
```

Comment:
I should have left a comment above the regex, but the line with /\D/g is removing all the non-digit characters from the hash (to make it easier to read over the phone). So if somehow you get a MD5 hash that has fewer than _RESET_CODE_LENGTH digits, then we have a problem. I presume that will basically never happen, but the case should still be handled.